man/stat161.1
.Dd January 5, 2015
.Dt STAT161 1
.Os System/161 2.x
.Sh NAME
.Nm stat161
.Nd System/161 status monitor
.Sh SYNOPSIS
.Nm stat161 Op Fl i Ar interval
.Sh DESCRIPTION
The
.Nm stat161
program prints hardware counters from a running System/161 process.
If no such process is running in the current directory,
.Nm stat161
waits until one starts, polling once per second.
If the System/161 process exits,
.Nm stat161
disconnects from it and waits for another to start.
.Pp
The
.Fl i Ar interval
option sets the reporting interval in seconds of virtual time; valid
values are between 0.00001 (10 microseconds) and 2.
The default is 0.2.
.Pp
The counters printed are as follows:
.Bl -tag -width kinsnsww -compact -offset indent
.It nsec
Virtual time in nanoseconds.
.It kinsns
Instructions retired in supervisor (kernel) mode.
.It uinsns
Instructions retired in user mode.
.It udud
Unproductive cycles spent in user mode.
(These represent instructions that were interrupted and aborted.)
.It idle
Cycles spent idle.
.It irqs
Number of interrupts taken.
.It exns
Nunmber of processor exceptions taken.
.It disk
Number of disk I/Os.
.It con
Number of console I/Os.
.It emu
Number of emufs I/Os.
.It net
Number of network I/Os.
.El
The values printed are totals across all CPUs.
.Sh FILES
.Bl -tag -width .sockets/meter -compact
.It Pa .sockets/meter
The socket used to read information.
.El
.Sh SEE ALSO
.Xr sys161 1
.Sh BUGS
The set of counters printed and the print interval are fixed.
The print interval depends on the internal timing of System/161 and
may thus vary depending on workload.


CHANGES
20160117 dholland	System/161 2.0.6 released.
20160117 dholland	util.c needs sys/types.h; from Nikhil Benesch.
20160116 dholland	Don't let the dropped input character messages flood.

20160114 dholland	System/161 2.0.5 released.
20160114 dholland	Fix a bug counting global-idle cycles.
20160114 dholland	Drop redundant columns in stat161 output.
20160114 dholland	Polish previous.
20160113 gwa		Make stat161's reporting interval adjustable.
20160113 gwa		Have stat161 report elapsed clock time per tick.
20160112 dholland	Print a message if the OS drops an input character.
20160108 dholland	Print why if disk geometry initializion fails.
20160105 dholland	Standardize and document the exit codes.
20160105 dholland	Allow K, M, etc. suffixes for disk/memory sizes.
20160105 dholland	Make the trace device reject (vs. ignore) invalid args.
20160105 dholland	Add -C option to supply device config arguments.

20151222 dholland	System/161 2.0.4 released.
20151222 dholland	Improve the way -X works.
20151222 dholland	Make trace161 -h output include the trace flag list.
20151222 dholland	Use shell printf instead of echo -n. Apparently Apple
........		still ships a broken sh whose echo prints literal "-n".
20151222 dholland	Apparently some Linux installs demand _GNU_SOURCE to
........		be allowed to see/use ftruncate.
20150808 dholland	Actually install the new prof.html. Sigh.
........		(issued as patch for 2.0.3)

20150804 dholland	System/161 2.0.3 released.
20150714 dholland	Fix the serial port FORCE logic meant to prevent
........		interrupt flapping. Thanks to Winnie Wu for a test
........		kernel and workload that reliably repeated the problem.
20150709 dholland	Recognize some obvious aliases for disk161 commands.
20150706 dholland	Add page on profiling to the manual.
20150706 dholland	Add dynamic profiling control to the trace device.
20150703 dholland	If profiling, write the profile out if we die().
20150703 dholland	Include seek times in disk activity trace output.

20150126 dholland	System/161 2.0.2 released.
20150126 dholland	Fix brown paper bag release. Sigh.
20150126 dholland	System/161 2.0.1 released.
20150121 dholland	Fix up some gdb threads issues appearing with gdb 7.8.
20150117 dholland	Update included .hgignore file, from Cary Gray.

20150115 dholland	System/161 2.0 released.
20150115 dholland	Add workaround to make disk rotdelay deterministic.
20150114 dholland	Fix stdint.h-related build problems on Linux.
20150114 dholland	Fix flock-related build problem on Linux.
20150109 dholland	Add software debugger requests to the trace device.
20150105 dholland	Make the testsuite run.
20150105 dholland	Add support for installing into a chroot/DESTDIR.
20150105 dholland	Add man pages for all executables.

20141209 dholland	System/161 1.99.10 released.
20141209 dholland	Add workaround for OpenBSD's srand/srandom breakage.

20140929 dholland	System/161 1.99.09 released.
20140929 dholland	Support the CACHE instruction as (for now) a nop.
20140929 dholland	Pretend to have 4K each L1 I/D caches.
20140929 dholland	Provide the MIPS32 config0 and config1 registers.
20140929 dholland	Edit/revise the processor docs.
20140929 dholland	Don't allow attempts to do r3k-style cache flushes.
20140929 dholland	Sort out status register definition going forward.
20140925 dholland	Fix address space ID printouts. From Peter Bailis.
20140925 dholland	Update disk config in sys161.conf.sample.
20140925 dholland	Added disk161 utility for manipulating disk images.
20140925 dholland	Warn if the size of a disk in sys161.conf is wrong.
20140925 dholland	Take the size of a disk from its image file size.
20140919 dholland	Fix build with gcc 4.8.
20140919 dholland	Provide flock compat for legacy OSes like Solaris.

20140821 dholland	System/161 1.99.08 released.
20140807 dholland	Use flock() on disk images to avoid accidents.
20140801 dholland	Improve gdb interface to treat cpus as "threads".
20140730 dholland	Rework tty handling. Now behaves when backgrounded.
20140730 dholland	Change sample disk image names to LHD{0,1}.conf.
20140730 dholland	Retune main loop now that timing works. Much faster.
20140730 dholland	Rework timing. Fix bogus large idle counts for real.
20140729 dholland	Rework main loop. Tidier and noticeably faster.
20140315 dholland	Handle EOF on input better, from Kevin Elphinstone.
20140314 dholland	Use -Wstrict-prototypes -Wmissing-prototypes with gcc.
20140123 dholland	Fix date on front page of docs.

20140123 dholland	System/161 1.99.07 released.
20140123 dholland	Add a -D option to enable a disk write doom counter.
20140122 dholland	Skip ELF note sections; from Vincent Danjean.
20131105 dholland	Change the processor ID value; discussion in docs.
20131105 dholland	Add mips coprocessor register banks ("selects").
20130530 dholland	Add a -X option to not hang waiting for the debugger.
20130530 dholland	Print more sensible cycle counts with stat161.
			(They are now totals across cpus, not averages.)
20130530 dholland	Add a -Z timeout option for monitoring user progress.
20130530 dholland	Count and report ll/sc and sync instructions.
20130530 dholland	Make configure script no longer leave "a.out" behind.
20130530 dholland	Purge ancient rcsid strings.
20130530 dholland	Accept the SYNC (memory barrier) instruction.
20130530 dholland	Have configure script check for _FILE_OFFSET_BITS=64.
20130422 dholland	Tidy up the disk initialization code further.
20130422 dholland	Fix missing initialization of dd_timedop in dev_disk.
			(Reported by Keno Fischer.)

20130422 dholland	System/161 1.99.06 released.
20130422 dholland	Update download URL link provided in the docs.
20130422 dholland	Fix one source of bogusly large idle cycle counts.
20130306 dholland	Hack around C99 strict-aliasing issues.
20130306 dholland	Fix build with gcc 4.5.
20130306 dholland	Initialize emufs create op's isdir return properly.
20130228 dholland	Doc fixes: typos, clarifications. Mention mips SYNC.

20110126 dholland	System/161 1.99.05 released.

stat/stat161.c
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <unistd.h>
#include <errno.h>
#include <signal.h>
#include "config.h"

#define MAXFIELDS	16
#define MAXHEADERLEN	16
#define PATH_SOCKET	".sockets/meter"
#define PROTO_VERSION	2

struct field {
	uint64_t lastval;
	unsigned width;
	char header[MAXHEADERLEN];
};

static struct field fields[MAXFIELDS];
static int nfields=0;
static int lines_since_header=100000;

static
void
reset(void)
{
	int i;

	for (i=0; i<MAXFIELDS; i++) {
		fields[i].lastval = 0;
		fields[i].width = 0;
		strcpy(fields[i].header, "??");
	}
	nfields = 0;
	lines_since_header = 10000;
}

static
uint64_t
getval(const char *s)
{
	/* We could (should?) use strtoull, but it's not always available. */
	uint64_t val = 0;
	unsigned i;

	for (i=0; s[i]; i++) {
		if (s[i]>='0' && s[i]<='9') {
			val = val*10 + (s[i]-'0');
		}
		else {
			break;
		}
	}
	return val;
}

static
void
setheaders(int nwords, char **words)
{
	int i;

	if (nwords > MAXFIELDS) {
		static int warned=0;
		if (!warned) {
			fprintf(stderr, "stat161: Too many data fields; "
				"increase MAXFIELDS and recompile\n");
			warned = 1;
		}
		nwords = MAXFIELDS;
	}

	for (i=0; i<nwords; i++) {
		if (strlen(words[i]) >= MAXHEADERLEN) {
			words[i][MAXHEADERLEN-1] = 0;
		}
		if (strlen(words[i]) > fields[i].width) {
			fields[i].width = strlen(words[i]);
		}
		strcpy(fields[i].header, words[i]);
	}

	nfields = nwords;
}

static
void
setwidths(int nwords, char **words)
{
	int i;
	for (i=0; i<nwords; i++) {
		unsigned width = atoi(words[i]);
		if (i < MAXFIELDS) {
			if (width > fields[i].width) {
				fields[i].width = width;
			}
		}
	}
}

static
void
showdata(int nwords, char **words)
{
	uint64_t val, printval;
	int i;
	char tmp[128];

	if (nwords != nfields) {
		printf("stat161: Invalid packet (wrong number of fields)\n");
		return;
	}

	if (lines_since_header > 21) {
		printf("\n");
		for (i=0; i<nfields; i++) {
			printf("%-*s ", fields[i].width, fields[i].header);
		}
		printf("\n");
		lines_since_header = 0;
	}

	for (i=0; i<nwords; i++) {
		val = getval(words[i]);
		printval = val - fields[i].lastval;
		fields[i].lastval = val;

		if (sizeof(uint64_t)==sizeof(unsigned long)) {
			snprintf(tmp, sizeof(tmp), "%lu", 
				 (unsigned long) printval);
		}
		else {
			snprintf(tmp, sizeof(tmp), "%llu", 
				 (unsigned long long) printval);
		}

		if (strlen(tmp) > fields[i].width) {
			fields[i].width = strlen(tmp);
		}

		printf("%-*s ", fields[i].width, tmp);
	}
	printf("\n");
	lines_since_header++;
}

static
void
processline(char *line)
{
	char *words[32];
	char *s;
	int nwords, i;

	nwords = 0;
	for (s = strtok(line, " \t\r\n"); s; s = strtok(NULL, " \t\r\n")) {
		if (nwords < 32) {
			words[nwords++] = s;
		}
	}

	if (nwords==0) {
		return;
	}
	if (!strcasecmp(words[0], "error")) {
		/* no actual useful messages */
		return;
	}
	else if (!strcasecmp(words[0], "hello") && nwords==2) {
		int ver = atoi(words[1]);
		if (ver != PROTO_VERSION) {
			fprintf(stderr, "stat161: Wrong protocol version %d\n",
				ver);
			exit(1);
		}
	}
	else if (!strcasecmp(words[0], "head") && nwords>1) {
		setheaders(nwords-1, words+1);
	}
	else if (!strcasecmp(words[0], "width") && nwords>1) {
		setwidths(nwords-1, words+1);
	}
	else if (!strcasecmp(words[0], "data") && nwords>1) {
		showdata(nwords-1, words+1);
	}
	else if (!strcasecmp(words[0], "ok")) {
		/* nothing */
	}
	else if (!strcasecmp(words[0], "bad")) {
		printf("stat161: sys161 error:");
		for (i=1; i<nwords; i++) {
			printf(" %s", words[i]);
		}
		printf("\n");
	}
	else {
		printf("stat161: Invalid packet (improper header)\n");
	}
}

static
void
dometer(int s, unsigned interval)
{
	char buf[4096], *x;
	size_t bufpos=0;
	int r;

	snprintf(buf, sizeof(buf), "INTERVAL %u\r\n", interval);
	r = write(s, buf, strlen(buf));
	if (r<0) {
		fprintf(stderr, "stat161: write: %s\n", strerror(errno));
		return;
	}

	while (1) {
		assert(bufpos<=sizeof(buf));
		if (bufpos==sizeof(buf)) {
			fprintf(stderr, "stat161: Input overrun\n");
			bufpos = 0;
		}
		r = read(s, buf+bufpos, sizeof(buf)-bufpos);
		if (r<0) {
			fprintf(stderr, "stat161: read: %s\n",
				strerror(errno));
			return;
		}
		else if (r==0) {
			return;
		}

		bufpos += r;
		
		while ((x = memchr(buf, '\n', bufpos))!=NULL) {
			*x = 0;
			x++;
			bufpos -= (x-buf);
			processline(buf);
			memmove(buf, x, bufpos);
		}
	}
}

static
int
opensock(void)
{
	struct sockaddr_un su;
	socklen_t len;
	int s;

	s = socket(AF_UNIX, SOCK_STREAM, 0);
	if (s<0) {
		fprintf(stderr, "stat161: socket: %s\n", strerror(errno));
		exit(1);
	}

	memset(&su, 0, sizeof(su));
	su.sun_family = AF_UNIX;
	snprintf(su.sun_path, sizeof(su.sun_path), "%s", PATH_SOCKET);
	len = SUN_LEN(&su);
#ifdef HAS_SUN_LEN
	su.sun_len = len;
#endif

	if (connect(s, (struct sockaddr *)&su, len)<0) {
		if (errno!=ECONNREFUSED && errno!=ENOENT) {
			fprintf(stderr, "stat161: connect: %s\n", 
				strerror(errno));
			sleep(5);
		}
		else {
			sleep(1);
		}
		close(s);
		return -1;
	}

	return s;
}

static
void
loop(unsigned interval)
{
	int s;

	printf("stat161: Connecting...\n");
	while (1) {
		s = opensock();
		if (s>=0) {
			printf("stat161: Connected.\n");
			reset();
			dometer(s, interval);
			close(s);
			printf("stat161: Disconnected.\n");
		}
	}
}

////////////////////////////////////////////////////////////
// main

static
void
usage(void)
{
	fprintf(stderr, "Usage: stat161 [-i interval]\n");
	fprintf(stderr, "   -i interval  Set sampling interval,"
		" in seconds (default 0.2)\n");
	exit(3);
}

int
main(int argc, char *argv[])
{
	int opt;
	double secs;

	/* Default sampling interval (0.2s). */
	unsigned nsecs = 200000000;

	while ((opt = getopt(argc, argv, "i:"))!=-1) {
		switch (opt) {
		    case 'i':
			secs = atof(optarg);
			if (secs > 2.0) {
				fprintf(stderr, "Sampling interval too large"
					" (max is 2)\n");
				exit(1);
			}
			nsecs = (unsigned)(secs * 1000000000);
			if (nsecs < 10000) {
				fprintf(stderr, "Sampling interval too small"
					" (min is 0.00001)\n");
				exit(1);
			}
			break;
		    default:
			usage();
		}
	}
	signal(SIGPIPE, SIG_IGN);
	loop(nsecs);
	return 0;
}
